File: textbook/python/01-dataflow/08_multi_tenant.py
# Copyright 2026 Terrene Foundation
# SPDX-License-Identifier: Apache-2.0
"""
# ════════════════════════════════════════════════════════════════════════
# Kailash SDK Textbook — DataFlow / Multi-Tenant Context Switching
# ════════════════════════════════════════════════════════════════════════
# OBJECTIVE: Isolate data using multi-tenant context switching with
#            sync and async context managers
# LEVEL: Advanced
# PARITY: Full — Rust has QueryInterceptor for tenancy
# VALIDATES: TenantContextSwitch, TenantInfo, get_current_tenant_id,
#            register_tenant, switch(), aswitch(), require_tenant,
#            deactivate_tenant, activate_tenant, get_stats()
#
# Run: uv run python textbook/python/01-dataflow/08_multi_tenant.py
# ════════════════════════════════════════════════════════════════════════
"""
from __future__ import annotations

import asyncio
import os
import tempfile

from dataflow import DataFlow, TenantContextSwitch, TenantInfo, get_current_tenant_id

# ── Setup ───────────────────────────────────────────────────────────

db_path = os.path.join(tempfile.gettempdir(), "textbook_dataflow_08.db")
db = DataFlow(database_url=f"sqlite:///{db_path}")


@db.model
class Project:
    name: str
    status: str = "active"


async def main():
    await db.initialize()

    # ── 1. ACCESS — db.tenant_context property ──────────────────────
    # TenantContextSwitch is accessed via db.tenant_context.
    # It provides sync and async context managers for safely switching
    # between tenant contexts with guaranteed isolation.

    ctx = db.tenant_context
    assert isinstance(ctx, TenantContextSwitch)
    print("TenantContextSwitch accessible via db.tenant_context")

    # ── 2. REGISTER tenants ─────────────────────────────────────────
    # Register tenants before switching to them. Each tenant has an id,
    # a human-readable name, and optional metadata.

    tenant_a = ctx.register_tenant("acme", "Acme Corporation", {"region": "APAC"})
    assert isinstance(tenant_a, TenantInfo)
    assert tenant_a.tenant_id == "acme"
    assert tenant_a.name == "Acme Corporation"
    assert tenant_a.metadata == {"region": "APAC"}
    assert tenant_a.active is True
    print(f"Registered tenant: {tenant_a.name}")

    tenant_b = ctx.register_tenant("globex", "Globex Inc.", {"region": "EU"})
    assert tenant_b.tenant_id == "globex"

    # Duplicate registration raises ValueError
    try:
        ctx.register_tenant("acme", "Acme Again")
        assert False, "Should raise ValueError"
    except ValueError as e:
        assert "already registered" in str(e)

    # Invalid tenant_id raises ValueError
    try:
        ctx.register_tenant("", "Empty ID")
        assert False, "Should raise ValueError"
    except ValueError:
        pass

    # ── 3. LIST and GET tenants ─────────────────────────────────────

    all_tenants = ctx.list_tenants()
    assert len(all_tenants) == 2
    assert all(isinstance(t, TenantInfo) for t in all_tenants)

    found = ctx.get_tenant("acme")
    assert found is not None
    assert found.tenant_id == "acme"

    not_found = ctx.get_tenant("nonexistent")
    assert not_found is None

    # ── 4. SWITCH — sync context manager ────────────────────────────
    # with ctx.switch(tenant_id) sets the tenant for the block.
    # Previous context is restored on exit, even on exception.

    # Before any switch, no tenant is active
    assert ctx.get_current_tenant() is None
    assert get_current_tenant_id() is None, "Module-level helper also returns None"

    with ctx.switch("acme") as tenant_info:
        assert isinstance(tenant_info, TenantInfo)
        assert tenant_info.tenant_id == "acme"

        # Inside the context, the current tenant is set
        assert ctx.get_current_tenant() == "acme"
        assert get_current_tenant_id() == "acme", "Module-level helper reads context"
        print(f"Inside sync switch: tenant={ctx.get_current_tenant()}")

    # After the context manager exits, the previous context is restored
    assert ctx.get_current_tenant() is None
    print("After sync switch: tenant restored to None")

    # ── 5. ASWITCH — async context manager ──────────────────────────
    # async with ctx.aswitch(tenant_id) — same semantics for async code.
    # contextvars automatically propagates context to async tasks.

    async with ctx.aswitch("globex") as tenant_info:
        assert tenant_info.tenant_id == "globex"
        assert ctx.get_current_tenant() == "globex"
        print(f"Inside async switch: tenant={ctx.get_current_tenant()}")

    assert ctx.get_current_tenant() is None
    print("After async switch: tenant restored to None")

    # ── 6. NESTED switches ──────────────────────────────────────────
    # Nested switches properly track and restore the previous context.

    with ctx.switch("acme"):
        assert ctx.get_current_tenant() == "acme"

        with ctx.switch("globex"):
            assert ctx.get_current_tenant() == "globex"

        # Inner switch exited — restored to "acme"
        assert ctx.get_current_tenant() == "acme"

    # Outer switch exited — restored to None
    assert ctx.get_current_tenant() is None
    print("Nested switches: proper context restoration verified")

    # ── 7. REQUIRE_TENANT — strict enforcement ──────────────────────
    # require_tenant() raises RuntimeError if no tenant context is active.
    # Useful for operations that must run within a tenant scope.

    try:
        ctx.require_tenant()
        assert False, "Should raise RuntimeError"
    except RuntimeError as e:
        assert "No tenant context is active" in str(e)

    with ctx.switch("acme"):
        tenant_id = ctx.require_tenant()
        assert tenant_id == "acme"

    # ── 8. DEACTIVATE and ACTIVATE tenants ──────────────────────────
    # Deactivated tenants remain registered but cannot be switched to.

    ctx.deactivate_tenant("globex")
    assert ctx.is_tenant_registered("globex") is True
    assert ctx.is_tenant_active("globex") is False

    # Switching to a deactivated tenant raises ValueError
    try:
        with ctx.switch("globex"):
            pass
        assert False, "Should raise ValueError"
    except ValueError as e:
        assert "not active" in str(e)

    # Reactivate the tenant
    ctx.activate_tenant("globex")
    assert ctx.is_tenant_active("globex") is True

    # Now switching works again
    with ctx.switch("globex"):
        assert ctx.get_current_tenant() == "globex"
    print("Deactivate/activate: tenant lifecycle verified")

    # ── 9. SWITCH to unregistered tenant ────────────────────────────
    # Raises ValueError with a helpful message listing available tenants.

    try:
        with ctx.switch("unknown-corp"):
            pass
        assert False, "Should raise ValueError"
    except ValueError as e:
        assert "not registered" in str(e)
        assert "Available tenants" in str(e)

    # ── 10. UNREGISTER a tenant ─────────────────────────────────────

    ctx.register_tenant("temp", "Temporary Tenant")
    assert ctx.is_tenant_registered("temp") is True

    ctx.unregister_tenant("temp")
    assert ctx.is_tenant_registered("temp") is False

    # Cannot unregister while active context
    with ctx.switch("acme"):
        try:
            ctx.unregister_tenant("acme")
            assert False, "Should raise ValueError"
        except ValueError as e:
            assert "active context" in str(e)

    # ── 11. STATS — context switching statistics ────────────────────

    stats = ctx.get_stats()
    assert isinstance(stats, dict)
    assert stats["total_tenants"] == 2, "Two registered tenants"
    assert stats["active_tenants"] == 2, "Both are active"
    assert stats["total_switches"] > 0, "Switches have been performed"
    assert stats["active_switches"] == 0, "No switch currently active"
    assert stats["current_tenant"] is None, "No current tenant outside switch"
    print(f"Stats: {stats['total_switches']} total switches performed")

    # ── 12. Context restoration on exception ────────────────────────
    # Even when an exception occurs, the previous context is restored.

    with ctx.switch("acme"):
        try:
            with ctx.switch("globex"):
                assert ctx.get_current_tenant() == "globex"
                raise RuntimeError("Simulated failure")
        except RuntimeError:
            pass

        # Restored to "acme" despite exception in inner block
        assert ctx.get_current_tenant() == "acme"

    assert ctx.get_current_tenant() is None
    print("Exception safety: context restored after error")

    # ── Cleanup ─────────────────────────────────────────────────────
    await db.stop()


asyncio.run(main())

try:
    os.unlink(db_path)
except OSError:
    pass

print("PASS: 01-dataflow/08_multi_tenant")

File: textbook/rust/01-dataflow/src/bin/08_multi_tenant.rs
// Copyright 2026 Terrene Foundation
// SPDX-License-Identifier: Apache-2.0
//! Kailash SDK Textbook -- DataFlow / Multi-Tenant
//!
//! OBJECTIVE: Use QueryInterceptor for automatic tenant-scoped database access.
//! LEVEL: Advanced
//! PARITY: Full -- Both SDKs provide TenantContext and QueryInterceptor that
//!         automatically inject WHERE tenant_id = ? into all queries.
//! VALIDATES: TenantContext, QueryInterceptor, TenantId, TenantContextMiddleware
//!
//! Run: cargo run -p tutorial-dataflow --bin 08_multi_tenant

use kailash_dataflow::prelude::*;
use kailash_value::{Value, ValueMap};
use std::sync::Arc;

#[tokio::main]
async fn main() {
    // ── 1. Multi-Tenant Architecture ──
    // In SaaS applications, each tenant's data must be isolated.
    // DataFlow's QueryInterceptor automatically adds tenant filters
    // to every query, preventing cross-tenant data leaks.

    let mut df = DataFlow::new("sqlite::memory:")
        .await
        .expect("connect");

    // ── 2. Model with Tenant Column ──
    // The tenant_id field is the partition key. The interceptor uses
    // this column to scope all queries.

    let model = ModelDefinition::new("Document", "documents")
        .field("id", FieldType::Integer, |f| f.primary_key())
        .field("tenant_id", FieldType::Text, |f| f.required())
        .field("title", FieldType::Text, |f| f.required())
        .field("content", FieldType::Text, |f| f)
        .auto_timestamps();

    df.register_model(model).expect("register");
    let express = df.express();

    // ── 3. Insert Data for Multiple Tenants ──

    let docs = vec![
        ("tenant_a", "Doc A1", "Content for tenant A"),
        ("tenant_a", "Doc A2", "More content for A"),
        ("tenant_b", "Doc B1", "Content for tenant B"),
        ("tenant_b", "Doc B2", "Tenant B's private doc"),
        ("tenant_b", "Doc B3", "Another B document"),
    ];

    for (tenant, title, content) in &docs {
        let mut data = ValueMap::new();
        data.insert(Arc::from("tenant_id"), Value::from(*tenant));
        data.insert(Arc::from("title"), Value::from(*title));
        data.insert(Arc::from("content"), Value::from(*content));
        express.create("Document", data).await.expect("insert");
    }

    // Without tenant filtering, all 5 documents are visible.
    let all = express
        .list("Document", ListOptions::default())
        .await
        .expect("list all");
    assert_eq!(all.len(), 5);

    // ── 4. TenantContext ──
    // TenantContext carries the current tenant ID for query scoping.

    let tenant_a = TenantContext::new(TenantId::from("tenant_a"));
    let tenant_b = TenantContext::new(TenantId::from("tenant_b"));

    assert_eq!(tenant_a.tenant_id().as_str(), "tenant_a");
    assert_eq!(tenant_b.tenant_id().as_str(), "tenant_b");

    // ── 5. QueryInterceptor ──
    // The interceptor modifies queries to add tenant_id = ? conditions.
    // This is typically applied as middleware in a Nexus server.

    // Manual filter equivalent (what the interceptor does automatically):
    let tenant_a_docs = express
        .list(
            "Document",
            ListOptions {
                filters: vec![FilterCondition::new(
                    "tenant_id",
                    FilterOp::Eq,
                    Value::from("tenant_a"),
                )],
                ..Default::default()
            },
        )
        .await
        .expect("list tenant_a");
    assert_eq!(tenant_a_docs.len(), 2);

    let tenant_b_docs = express
        .list(
            "Document",
            ListOptions {
                filters: vec![FilterCondition::new(
                    "tenant_id",
                    FilterOp::Eq,
                    Value::from("tenant_b"),
                )],
                ..Default::default()
            },
        )
        .await
        .expect("list tenant_b");
    assert_eq!(tenant_b_docs.len(), 3);

    // ── 6. Tenant Isolation Guarantee ──
    // Each tenant sees ONLY their own data. No cross-tenant leaks.

    for doc in &tenant_a_docs {
        assert_eq!(
            doc.get("tenant_id" as &str),
            Some(&Value::from("tenant_a")),
            "tenant_a query should only return tenant_a docs"
        );
    }

    for doc in &tenant_b_docs {
        assert_eq!(
            doc.get("tenant_id" as &str),
            Some(&Value::from("tenant_b")),
            "tenant_b query should only return tenant_b docs"
        );
    }

    println!("PASS: 01-dataflow/08_multi_tenant");
}
